Add option to specify the location of the tests directory

jwasinger opened the pull request:
Sometimes, it is useful to have tests reside in a location other than the default.

holgerd77 commented:
Can you rebase and fix the conflicts?

holgerd77 commented:
No library in in ``ethereumjs`` repos is [directly using](https://github.com/search?q=org%3Aethereumjs+getTests&type=Code) the ``getTests``  method except ``ethereumjs-tx``, which is actually using an extremely outdated and fixed ``0.0.1`` version, so I think you can change the ``getTests`` API and add the parameter in a more intuitive place preceding ``testType`` (I actually also adopted the``getTests`` API with the PR for choosing the ``testDir`` and I'm generally not sure about the state of ``getTests``, maybe this function shouldn't be exported at all?).

Generally we then have three CL options for combining the path together, ``testsDir``, ``testType`` and ``testDir`` (``dir`` as CL option). I am personally fine with the three options (otherwise it also would create too much fuzz changing the usage of this library). But maybe rename your option to ``testsPath`` at least, otherwise this will cause too much confusion.

holgerd77 commented on index.js:
```diff
@@ -66,7 +66,11 @@ exports.getTestsFromArgs = function (testType, onFile, args = {}) {
     }
   }
-  return getTests(testType, onFile, fileFilter, skipFn)
+  if (args.testsDir) {
```
Can you stay in the logic used for the other arguments for calling ``getTests`` instead of this call switch, e.g.

```
let testDir, fileFilter, excludeDir, skipFn

if (testType === 'VMTests') {
    skipFn = (name) => {
      return skipTest(name, args.skipVM)
    }
}
  
if (args.dir) {   
  testDir = args.dir
}
```

Not sure if this is the most elegant way of doing this but at least then it's consistent.

jwasinger replied:
Sure, this is reasonable.

holgerd77 commented:
(sorry, this is getting a bit fuzzy for a change maybe not worth all the excitement. Another argument for the review process. My PR was the only one I merged without reviewing, if I would have done we would probably had figured out this stuff before...)

jwasinger commented:
>Generally we then have three CL options for combining the path together, testsDir, testType and testDir (dir as CL option). I am personally fine with the three options (otherwise it also would create too much fuzz changing the usage of this library). But maybe rename your option to testsPath at least, otherwise this will cause too much confusion.

Agreed.

>No library in in ethereumjs repos is directly using the getTests method except ethereumjs-tx, which is actually using an extremely outdated and fixed 0.0.1 version, so I think you can change the getTests API and add the parameter in a more intuitive place preceding testType (I actually also adopted thegetTests API with the PR for choosing the testDir and I'm generally not sure about the state of getTests, maybe this function shouldn't be exported at all?).

This function is called within `getTestsFromArgs`. So it is used.

holgerd77 commented:
I mean API usage from the outside, everything which happens inside the library we have under control.

jwasinger committed:
add option to specify the path to tests directory as a command line argument

jwasinger commented:
>No library in in ethereumjs repos is directly using the getTests method except ethereumjs-tx, which is actually using an extremely outdated and fixed 0.0.1 version, so I think you can change the getTests API and add the parameter in a more intuitive place preceding testType (I actually also adopted thegetTests API with the PR for choosing the testDir and I'm generally not sure about the state of getTests, maybe this function shouldn't be exported at all?).

If we are concerned about not breaking other projects that depend on `ethereumjs-testing`, we need to preserve the order of function arguments in exposed functions.

I changed the order of the arguments in `getTests` so that it is backwards compatible with any existing code that uses it.

holgerd77 commented:
@jwasinger Please add the new ``testsPath`` arg also to the API description/docs in the README file.

One general thing to consider: is this change really necessary? Another option is to just symlink into the desired ``tests`` directory. Has this major drawbacks or is less convenient then the solution you want to implement here? Don't want to start a big discussion here though, just wanted to throw this once in the ring. If you say, "would be nice to have", this would be fine.

holgerd77 committed:
Added testsPath arg description to README

holgerd77 reviewed (approved):
Looks good to me
